server/irods.go

```diff
@@ -214,25 +243,117 @@ func isReadableByUser(logger zerolog.Logger, filesystem *ifs.FileSystem,
 			effectiveUserZone = userZone
 		}
 
-		if effectiveUserZone == pathZone &&
-			ac.UserName == userName &&
-			ac.AccessLevel == types.IRODSAccessLevelReadObject {
+		hasRead := ac.AccessLevel == types.IRODSAccessLevelReadObject
+		hasOwn := ac.AccessLevel == types.IRODSAccessLevelOwner
+
+		// There is permission directly for the user
+		if ac.UserType == types.IRODSUserRodsUser || ac.UserType == types.IRODSUserRodsAdmin {
+			if effectiveUserZone == pathZone && ac.UserName == userName && (hasRead || hasOwn) {
+				logger.Trace().
+					Str("path", rodsPath).
+					Str("user", userName).
+					Str("zone", userZone).
+					Str("effective_zone", effectiveUserZone).
+					Str("ac_user", ac.UserName).
+					Str("ac_level", string(ac.AccessLevel)).
+					Bool("read", hasRead).
+					Bool("own", hasOwn).
+					Msg("User access found")
+
+				return true, nil
+			}
+
 			logger.Trace().
 				Str("path", rodsPath).
 				Str("user", userName).
 				Str("zone", userZone).
-				Msg("Read access found")
+				Str("ac_user", ac.UserName).
+				Str("effective_zone", effectiveUserZone).
+				Str("path_zone", pathZone).
+				Str("ac_level", string(ac.AccessLevel)).
+				Bool("read", hasRead).
+				Bool("own", hasOwn).
+				Msg("User read access not found")
+		}
 
-			return true, nil
+		// There is permission for a group the user is in
+		if ac.UserType == types.IRODSUserRodsGroup {
+			// Check if user in the group of this AC (ac.UserName is the name of the AC's group, unfortunately)
+			_, userInGroup := userGroupLookup[ac.UserName]
+
+			if effectiveUserZone == pathZone && userInGroup && (hasRead || hasOwn) {
+				logger.Trace().
+					Str("path", rodsPath).
+					Str("user", userName).
+					Str("zone", userZone).
+					Str("effective_zone", effectiveUserZone).
+					Str("ac_user", ac.UserName).
+					Str("ac_level", string(ac.AccessLevel)).
+					Bool("read", hasRead).
+					Bool("own", hasOwn).
+					Bool("user_in_group", userInGroup).
+					Msg("Group access found")
+
+				return true, nil
+			}
 		}
 	}
 
 	logger.Trace().
 		Str("path", rodsPath).
 		Str("user", userName).
 		Str("zone", userZone).
-		Msg("Read access not found")
+		Msg("No access found")
+
+	return false, nil
+}
+
+func UserInGroup(logger zerolog.Logger, filesystem *ifs.FileSystem,
+	userName string, userZone string, groupName string) (_ bool, err error) {
+	var groups []*types.IRODSUser
+	if groups, err = filesystem.ListUserGroups(userName); err != nil {
+		return false, err
+	}
+
+	for _, group := range groups {
+		logger.Trace().
+			Str("user", userName).
+			Str("zone", userZone).
+			Str("group", group.Name).
+			Msg("Checking user group")
+
+		if group.Zone == "" {
+			return group.Name == groupName, nil
+		}
+
+		return group.Zone == userZone && group.Name == groupName, nil
+	}
```

Comment:
This looks like a bug; returning the first result, rather than scanning the whole list.

Reply:
Added a fix